src/Squirrel.CommandLine/Deployment/S3Commands.cs

```diff
@@ -0,0 +1,132 @@
+                        result.ErrorMessage = "A region value is required";
+                    }
+                }
+            });
+            Add(Region);
+
+            Endpoint = new Option<string>("--endpoint", "Custom service url (backblaze, digital ocean, etc).") {
+                ArgumentHelpName = "URL"
+            };
+            Add(Endpoint);
+
+            Bucket = new Option<string>("--bucket", "Name of the S3 bucket.") {
+                ArgumentHelpName = "NAME",
+                IsRequired = true
+            };
+            Add(Bucket);
+
+            PathPrefix = new Option<string>("--pathPrefix", "A sub-folder used for files in the bucket, for creating release channels (eg. 'stable' or 'dev').") {
+                ArgumentHelpName = "PREFIX"
+            };
+            Add(PathPrefix);
+
+            this.AreMutuallyExclusive(Region, Endpoint)
+                .AtLeastOneRequired(Region, Endpoint);
+        }
+
+        private protected void SetOptionsValues(InvocationContext context, SyncS3Options options)
+        {
+            base.SetOptionsValues(context, options);
+            options.keyId = context.ParseResult.GetValueForOption(KeyId);
+            options.secret = context.ParseResult.GetValueForOption(Secret);
+            options.region = context.ParseResult.GetValueForOption(Region);
+            options.endpoint = context.ParseResult.GetValueForOption(Endpoint);
+            options.bucket = context.ParseResult.GetValueForOption(Bucket);
+            options.pathPrefix = context.ParseResult.GetValueForOption(PathPrefix);
+        }
+    }
+
+    public class S3DownloadCommand : S3BaseCommand
+    {
+        public S3DownloadCommand()
+            : base("s3", "Download latest release from an S3 bucket.")
+        {
+            this.SetHandler(Execute);
+        }
+
+        private async Task Execute(InvocationContext context)
+        {
+            SyncS3Options options = new();
+            SetOptionsValues(context, options);
+            await new S3Repository(options).DownloadRecentPackages();
+        }
+    }
+
+    public class S3UploadCommand : S3BaseCommand
+    {
+        public Option<bool> Overwrite { get; }
+        public Option<int> KeepMaxReleases { get; }
+
+        public S3UploadCommand()
+            : base("s3", "Upload releases to an S3 bucket.")
+        {
+            Overwrite = new Option<bool>("--overwrite", "Replace remote files if local files have changed.");
+            Add(Overwrite);
+
+            KeepMaxReleases = new Option<int>("--keepMaxReleases", "Apply a retention policy which keeps only the specified number of old versions in remote source.") {
+                ArgumentHelpName = "NUMBER"
+            };
+            Add(KeepMaxReleases);
+
+            this.SetHandler(Execute);
+        }
+
+        //Intentionally hiding base member
+        private protected new void SetOptionsValues(InvocationContext context, SyncS3Options options)
```

Comment:
Why are these `new` instead of `override`?

Reply:
For some reason I was thinking that I needed to do this to handle the duplicate signatures. I have fixed this.